### PR TITLE
Replace find-ts-imports with swc ecmascript parser

### PR DESCRIPTION
**Pull request by flickyiyo**

### CODE REVIEW

File: src/walk_dirs.rs [hunk 1]
```diff
@@ -144,13 +155,120 @@ pub fn discover_fences_and_files(start_path: &str) -> Vec<WalkFileData> {
...
+        None
+      }).collect();
+
+      if let Some(current_set) = imports_map.get(&import.source.value.to_string()) {
+        if let  Some(current_set) = current_set {
+          let mut new_set: HashSet<String> = HashSet::from_iter(current_set.iter().map(|v| v).cloned());
+          for val in set {
+            new_set.insert(val);
+          }
+          imports_map.insert(import.source.value.to_string(), Some(new_set.to_owned()));
```
Comment (Adjective-Object) [inline src/walk_dirs.rs:190]: you should be able to mutate this instead of copying and overwriting, I think you need the map to be 
`HashMap<String, Option<mut HashSet<String>>>`?

Comment (Adjective-Object) [reply src/walk_dirs.rs:190]: No worries if you grind your head against it too long and get stuck, we can always leave a TODO and file a github issue then revisit later

File: src/walk_dirs.rs [hunk 2]
```diff
@@ -144,13 +155,120 @@ pub fn discover_fences_and_files(start_path: &str) -> Vec<WalkFileData> {
...
+
+  for e in parser.take_errors() {
+    e.into_diagnostic(&handler).emit();
+  }
+
+
+
+  let ts_module = parser.parse_typescript_module()
+    .map_err(|e| e.into_diagnostic(&handler).emit())
+    .expect("Failed to parse module.");
```
Comment (Adjective-Object) [inline src/walk_dirs.rs:240]: You should handle these errors and return Result<> here, since the program shouldn't _crash_ when it encounters invalid input

Comment (Adjective-Object) [reply src/walk_dirs.rs:240]: Specifically, it's expected for devs to run it against unparseable code and we should give them a human readable error (and a partial fence evaluation) rather than crashing with a stack trace when that happens

File: src/walk_dirs.rs [hunk 3]
```diff
@@ -308,4 +423,31 @@ mod test {
       discovered
     );
   }
+
+  #[test]
+  fn tesT_get_imports_from_file() {
```
Comment (Adjective-Object) [inline src/walk_dirs.rs:428]: tesT -> test

**Review by Adjective-Object — CHANGES_REQUESTED** [review]
Looking good so far! Just a few cleanliness things

File: src/walk_dirs.rs (hunk 1 above)
Comment (flickyiyo) [reply src/walk_dirs.rs]: Added a TODO comment on this

File: src/file_import_retriever.rs [hunk 4]
```diff
@@ -0,0 +1,162 @@
+use std::collections::{HashMap, HashSet};
```
Comment (Adjective-Object) [inline src/file_import_retriever.rs:1]: Since this file's main export is 2 functions which are verbs (get_imports_from_file, get_imports_map), it should be named a verb as well, maybe something like `get_imports` rather than `file_import_retriever`.

`good-fences_runner` is a noun because its main export is a struct + that struct's impl.

File: src/file_import_retriever.rs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    importer_file_path: &PathBuf,
+) -> HashMap<String, Option<HashSet<String>>> {
+    let mut imports_map: HashMap<String, Option<HashSet<String>>> = HashMap::new();
+
+    imports.iter().for_each(|import| {
+        let set: HashSet<String> = import
+					.specifiers
+					.iter()
+					.filter_map(|spec| -> Option<String> {
+						let file_text =std::fs::read(importer_file_path).expect(&format!("error opening source file \"{:?}\"", importer_file_path));
```
Comment (Adjective-Object) [inline src/file_import_retriever.rs:75]: can you run `cargo fmt` before merging? just to clean up this weird whitespace. If it gets messy / has unrelated changes it could be a follow-up PR.

File: src/file_import_retriever.rs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+						if let Some(default) = spec.as_default() {
+							let text = get_string_of_span(&file_text, &default.span);
+							return Some(text);
```
Comment (Adjective-Object) [inline src/file_import_retriever.rs:78]: nit: prefer `return Some(get_string_of_span(&file_text, &default.span))`

File: src/file_import_retriever.rs [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+						}
+						if let Some(named) = spec.as_named() {
+							let text = get_string_of_span(&file_text, &named.span);
+							return Some(text);
```
Comment (Adjective-Object) [inline src/file_import_retriever.rs:82]: nit: prefer `return Some(get_string_of_span(&file_text, &named.span))`

File: src/file_import_retriever.rs [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+						} else {
+							imports_map.insert(import.source.value.to_string(), Some(set));
+						}
+					}
+        }
+    });
+    imports_map
+}
+
+pub fn create_lexer<'a>(fm: &'a swc_common::SourceFile) -> Lexer<'a, StringInput<'a>> {
```
Comment (Adjective-Object) [inline src/file_import_retriever.rs:109]: does this need to be public?

File: src/file_import_retriever.rs [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+                }
+            }
+            None
+        })
+        .collect();
+
+    return imports;
+}
+
+pub fn get_string_of_span<'a>(file_text: &'a Vec<u8>, span: &'a swc_common::Span) -> String {
```
Comment (Adjective-Object) [inline src/file_import_retriever.rs:59]: does this need to be public?

File: src/walk_dirs.rs [hunk 10]
```diff
@@ -1,13 +1,12 @@
...
-use find_ts_imports::{parse_source_file_imports, SourceFileImportData};
+use crate::file_import_retriever;
 use jwalk::WalkDirGeneric;
 use relative_path::RelativePath;
 use serde::Deserialize;
-use std::collections::HashSet;
+use std::collections::{HashSet, HashMap};
 use std::iter::FromIterator;
 use std::path::{Path, PathBuf};
+use file_import_retriever::{get_imports_from_file, get_imports_map};
```
Comment (Adjective-Object) [inline src/walk_dirs.rs:9]: This could be condensed with the import on like 2 to `crate::file_import_retriever::{...}`.

Comment (Adjective-Object) [reply src/walk_dirs.rs:9]: Though to be honest the imports are a mess to start with, apologies for that.

File: src/walk_dirs.rs [hunk 11]
```diff
@@ -144,11 +146,12 @@ pub fn discover_fences_and_files(start_path: &str) -> Vec<WalkFileData> {
     .collect();
 }
 
+
+
```
Comment (Adjective-Object) [inline src/walk_dirs.rs:150]: unnecessary whitespace changes

File: src/walk_dirs.rs [hunk 12]
```diff
@@ -119,9 +118,12 @@ pub fn discover_fences_and_files(start_path: &str) -> Vec<WalkFileData> {
                   let _working_dir_path: &Path = &WORKING_DIR_PATH;
                   let source_file_path = RelativePath::from_path(&file_path);
 
+                  let imports = get_imports_from_file(&file_path);
+                  let imports_map = get_imports_map(&imports, &file_path);
```
Comment (Adjective-Object) [inline src/walk_dirs.rs:122]: These functions are only ever called in succession -- I think they should be made private and instead the module should only export `get_imports_map_from_file`.

File: src/file_import_retriever.rs [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+    let mut parser = Parser::new_from(capturing);
+
+    for e in parser.take_errors() {
+        e.into_diagnostic(&handler).emit();
+    }
+
+    let ts_module = parser
+        .parse_typescript_module()
+        .map_err(|e| e.into_diagnostic(&handler).emit())
+        .expect("Failed to parse module.");
```
Comment (Adjective-Object) [inline src/file_import_retriever.rs:35]: You should handle this error and return an error type here. You could do it in this change or a follow-up PR, but please do it eventually 🙏🥺

File: src/walk_dirs.rs (hunk 11 above)
Comment (flickyiyo) [reply src/walk_dirs.rs]: Does `cargo fmt` resolve this?

Comment (Adjective-Object) [reply src/walk_dirs.rs:149]: yep

**Review by Adjective-Object — APPROVED** [review]
lgtm!
